src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp

```diff
@@ -64,48 +71,101 @@ class WaveClipTrimHandle::AdjustBorder final : public WaveClipTrimHandle::ClipTr
       }
    }
 
+   //Search for a good snap points among all tracks, including
+   //the one to which adjusted clip belongs to, but not counting
+   //the borders of adjusted clip
+   static SnapPointArray FindSnapPoints(
+      const WaveTrack* currentTrack, 
+      WaveClip* adjustedClip,
+      const std::pair<double, double> range)
+   {
+      SnapPointArray result;
+
+      auto addSnapPoint = [&](double t, const Track* track)
+      {
+         if(t > range.second || t < range.first)
+            return;
+
+         for(const auto& snapPoint : result)
+            if(snapPoint.t == t)
+               return;
+         result.emplace_back(t, track);
+      };
+
+      if(const auto trackList = currentTrack->GetOwner())
+      {
+         for(const auto track : trackList->Any())
+         {
+            const auto isSameTrack = (track == currentTrack) ||
+               (track->GetLinkType() == Track::LinkType::Aligned && *trackList->FindLeader(currentTrack) == track) ||
+               (currentTrack->GetLinkType() == Track::LinkType::Aligned && *trackList->FindLeader(track) == currentTrack);
+            for(const auto& interval : track->GetIntervals())
+            {
+               if(isSameTrack)
+               {
+                  auto waveTrackIntervalData = dynamic_cast<WaveTrack::IntervalData*>(interval.Extra());
```

Comment:
Just a discussion...

I would prefer `static_cast` myself here...  But actually, I'd rather see `std::any_cast`.  The Extra information in IntervalData looks like a place I could rewrite things with `std::any` now that we use C++17.  I think `any_cast` makes a simpler check than `dynamic_cast`.  Just equality of a statically determined `typeid` of the template parameter with the dynamically determined `typeid` in the `any`.

Reply:
I've never seen much performance impact from using `dynamic_cast`. It can be slow in tight loops with deep hierarchies, but that's not the case here.

Reply:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Pro-type-downcast

I agree that sometimes we can justify using `static_cast` instead of `dynamic_cast` if we want to get as much performance as possible, and we do guarantee that the specified type is the only possible option here. But one of the reasons why this part of the code was rewritten is that we may want to add more interval types to a wave track in the future, wasn't it?

Reply:
`std::any` has costs attached. It's not a `zero-cost` abstraction in general, although compilers may implement a fast-pass for small objects.